### DISCUSSION TITLE
Sharing my ShellGPT Aliases/Functions for Bash and Zsh

### DISCUSSION THREAD

**Discussion by WolframRavenwolf** [TheR1D/shell_gpt#409] [Show and tell]
I use ShellGPT daily but it could be more user-friendly (can't remember all the options), so I've created some **[ShellGPT Aliases/Functions for Bash and Zsh](https://gist.github.com/WolframRavenwolf/fb68cdda54193b154a3016d2626e152b)** to make it much easier (i. e. intuitive) to use. I'm sharing them here with other users, and if you'd like to integrate or adapt them into ShellGPT itself, @TheR1D, I'd be happy to contribute this.

## Setup

Just put the contents of this **[gist](https://gist.github.com/WolframRavenwolf/fb68cdda54193b154a3016d2626e152b)** into your `~/.bash_aliases` or `~/.oh-my-zsh/custom/aliases.zsh` or any other location where you keep your shell aliases/functions. Log out and back in or `source` the aliases file to load them, as usual.

## Features

The most important features are the shortcut functions - they're just a letter and a colon (so super easy to type, and no conflict with existing commands):

- `q:` chat ("**q**uery")
- `s:` **s**hell
- `d:` **d**escribe-shell
- `c:` **c**ode

Use **uppercase** letters for a bigger and better (and more expensive) model, switching from GPT-3.5 (pre-v1.0) or GPT-4-Turbo (v1.0) to **GPT-4**. **Double uppercase** letters are supposed to expand context by switching to **GPT-4-32K** (although that model isn't available to me right now?).

If you **don't** append your input to the shortcut, it will open **REPL** mode, so you see the chat history and get an interactive session.

It keeps a **chat cache** for each letter individually, so you can do follow-up questions and freely switch between command and REPL modes.

When you follow the shortcut with a **minus/dash**, it **clears the chat cache** of only that function (chat/shell/describe/code). **Double minuses/dashes** **clear the request cache**. **Triple minuses/dashes** **clear both caches** (the chat cache of only that function and the shared request cache).

Besides these main functions, there are also a bunch of aliases like `sgpt-chat` for `sgpt --repl chat`, but I rarely use those as the letter shortcut functions explained above are so much more useful. There are also clear commands, and `sgpt-clear-chats` could be useful when you want to clear all chat histories at once, as the letter shortcuts only clear their own chat history.

Also of note is that the cache paths are taken from the `.sgptrc` config file, so it works with custom paths as well. I prefer my caches to be inside my own home directory's `.cache` folder instead of the system-wide `/tmp` folder (and actually think that should be the standard, following the XDG Base Directory Specification and would be more organized and secure that way – but that's just a suggestion).

## Examples

Here are the same examples from the [Usage](https://github.com/TheR1D/shell_gpt#usage) section of the README, using my shell aliases/functions instead:

```sh
# Simple queries
q: nginx default config file location
Q: mass of sun # using GPT-4
QQ: 1 hour and 30 minutes to seconds # using GPT-4-32K

# Summarization and analyzing
git diff | q: Generate git commit message, for my changes
docker logs -n 20 container_name | Q: check logs, find errors, provide possible solutions # using GPT-4

# Shell commands
s: make all files in current directory read only
S: update my system # using GPT-4
cat data.json | s: curl localhost with provided json
s: using ffmpeg combine multiple videos into one without audio. Video file names: $(ls -m)

# Generating code
c: Solve classic fizz buzz problem using Python
C: solve classic fizz buzz problem using Python > fizz_buzz.py # using GPT-4
cat fizz_buzz.py | CC: Generate comments for each line of my code # using GPT-4-32K

# Chat Mode
q: please remember my favorite number: 4
q: "what would be my favorite number + 4?" # quoting required because of shell special character "?"
c: make an example request to localhost using Python
C: add caching # using GPT-4
s: What are the files in this directory"?" # it's enough to quote just shell special character(s)
s: Sort them by name
s: Concatenate them using FFMPEG
S: Convert the resulting file into an MP3 # using GPT-4

# REPL Mode
q:
s:
d:
c:
Q: # using GPT-4
S: # using GPT-4
D: # using GPT-4
C: # using GPT-4
QQ: # using GPT-4-32K
SS: # using GPT-4-32K
DD: # using GPT-4-32K
CC: # using GPT-4-32K

# Clear chat cache
q:- # clear chat cache
s:- # clear shell cache
d:- # clear describe-shell cache
c:- # clear code cache

# Clear request cache
q:-- # clear request cache (same for all letters)

# Clear all caches
q:-- # clear all caches (same for all letters)
```

Hope this is helpful to any of you!

**Comment by ristomatti** [comment]
Very intuitive and clever! I would have never thought `:` is a valid character for shell function names. Although maybe I should have as it was ~10y ago I came up with this simple hack:

```sh
alias ..='cd ../..'
alias ...='cd ../../..'
alias ....='cd ../../../..'
alias .....='cd ../../../../..'
```

**Comment by kuldeepluvani** [comment]
@WolframRavenwolf Your GIST's are dead line. 💀 Can you please repost?

**Reply by WolframRavenwolf** [reply]
@kuldeepluvani Whoops! Fixed. ✅ Thanks for letting me know. Enjoy the aliases!
